File: frontend/src/service/jniService.ts
import apiClient from "../api/client";
import type {
    BridgeConnectionForm,
    BridgeConnectionState,
    CalibrationRequest,
    CalibrationGlobalSettingsRecord,
    CalibrationGlobalSettingsRequest,
    CalibrationPreviewRecord,
    CalibrationSessionRecord,
    FpgaPayloadPixelDataRecord,
    FpgaPayloadPixelDataRequest,
    ImageFrameRecord,
    ImagePixelDataRecord,
    ImagePixelDataRequest,
    MultiFrameAnalysisRecord,
    MultiFrameAnalysisRequest,
    SpectrumExtractionRecord,
    SpectrumExtractionRequest,
    TriggerCaptureOptions,
    TriggerCaptureResponse,
} from "../types/jni";

type CalibrationType = "DARK" | "FLAT" | "HDR_DARK" | "HDR_FLAT";

export const jniApi = {
    getState: () => apiClient.get<unknown, BridgeConnectionState>("/jni/state"),
    connect: (payload: BridgeConnectionForm) =>
        apiClient.post<unknown, BridgeConnectionState>("/jni/connect", payload),
    disconnect: () => apiClient.post<unknown, BridgeConnectionState>("/jni/disconnect"),
    sendReset: () => apiClient.post<unknown, string>("/jni/commands/reset"),
    sendTriggerOnce: (payload?: TriggerCaptureOptions) =>
        apiClient.post<unknown, TriggerCaptureResponse>(
            "/jni/commands/trigger-once",
            payload ?? { autoProcess: false },
        ),
    sendQueryStatus: () => apiClient.post<unknown, string>("/jni/commands/query-status"),
    sendFullConfig: (configBytes: number[]) =>
        apiClient.post<unknown, string>("/jni/commands/full-config", { configBytes }),
    listImages: () => apiClient.get<unknown, ImageFrameRecord[]>("/jni/images"),
    listHdrImages: () => apiClient.get<unknown, ImageFrameRecord[]>("/jni/hdr/images"),
    listHdrDarkImages: () => apiClient.get<unknown, ImageFrameRecord[]>("/jni/hdr-dark/images"),
    listHdrFlatImages: () => apiClient.get<unknown, ImageFrameRecord[]>("/jni/hdr-flat/images"),
    processImage: (imageId: number) => apiClient.post<unknown, ImageFrameRecord>(`/jni/images/${imageId}/process`),
    getImagePixels: (imageId: number, payload?: ImagePixelDataRequest) =>
        apiClient.get<unknown, ImagePixelDataRecord>(`/jni/images/${imageId}/pixels`, {
            params: payload,
        }),
    getFpgaPayloadPixels: (imageId: number, payload?: FpgaPayloadPixelDataRequest) =>
        apiClient.get<unknown, FpgaPayloadPixelDataRecord>(`/jni/images/${imageId}/fpga-payload`, {
            params: payload,
        }),
    getLatestSpectrum: (imageId: number) =>
        apiClient.get<unknown, SpectrumExtractionRecord | null>(`/jni/images/${imageId}/spectrum`),
    extractSpectrum: (imageId: number, payload?: SpectrumExtractionRequest) =>
        apiClient.post<unknown, SpectrumExtractionRecord>(`/jni/images/${imageId}/spectrum/extract`, payload ?? {}),
    deleteImage: (imageId: number) => apiClient.delete<unknown, boolean>(`/jni/images/${imageId}`),
    clearImages: () => apiClient.delete<unknown, number>("/jni/images"),
    analyzeMultiFrame: (payload: MultiFrameAnalysisRequest) =>
        apiClient.post<unknown, MultiFrameAnalysisRecord>("/jni/images/multi-frame/analyze", payload),
    simulateCalibration: (type: CalibrationType, payload?: CalibrationRequest) =>
        apiClient.post<unknown, CalibrationSessionRecord>(`/jni/calibrations/${type}/simulate`, payload ?? {}),
    buildCalibrationFromImages: (type: CalibrationType, payload: CalibrationRequest) =>
        apiClient.post<unknown, CalibrationSessionRecord>(`/jni/calibrations/${type}/from-images`, payload),
    listCalibrations: (type?: CalibrationType) =>
        apiClient.get<unknown, CalibrationSessionRecord[]>(
            "/jni/calibrations",
            type ? { params: { type } } : undefined,
        ),
    getCalibration: (sessionId: number) =>
        apiClient.get<unknown, CalibrationSessionRecord>(`/jni/calibrations/${sessionId}`),
    deleteCalibration: (sessionId: number) =>
        apiClient.delete<unknown, boolean>(`/jni/calibrations/${sessionId}`),
    listCalibrationPreviews: (sessionId: number, limit = 0) =>
        apiClient.get<unknown, CalibrationPreviewRecord[]>(`/jni/calibrations/${sessionId}/previews`, {
            params: { limit },
        }),
    getCalibrationReferencePreview: (sessionId: number) =>
        apiClient.get<unknown, CalibrationPreviewRecord | null>(`/jni/calibrations/${sessionId}/reference-preview`),
    listCalibrationReferencePreviews: (sessionId: number) =>
        apiClient.get<unknown, CalibrationPreviewRecord[]>(`/jni/calibrations/${sessionId}/reference-previews`),
    getCalibrationGlobalSettings: () =>
        apiClient.get<unknown, CalibrationGlobalSettingsRecord>("/jni/calibrations/settings"),
    updateCalibrationGlobalSettings: (payload: CalibrationGlobalSettingsRequest) =>
        apiClient.put<unknown, CalibrationGlobalSettingsRecord>("/jni/calibrations/settings", payload),
};

File: springboot-jni/src/main/java/springbootjni/service/JNIService.java
package springbootjni.service;

import springbootjni.dto.jni.BridgeConnectRequest;
import springbootjni.dto.jni.BridgeStateResponse;
import springbootjni.dto.jni.CalibrationRequest;
import springbootjni.dto.jni.CalibrationGlobalSettingsRequest;
import springbootjni.dto.jni.CalibrationGlobalSettingsResponse;
import springbootjni.dto.jni.CalibrationPreviewResponse;
import springbootjni.dto.jni.CalibrationSessionResponse;
import springbootjni.dto.jni.FpgaPayloadPixelDataResponse;
import springbootjni.dto.jni.ImageFrameResponse;
import springbootjni.dto.jni.ImagePixelDataResponse;
import springbootjni.dto.jni.MultiFrameAnalysisRequest;
import springbootjni.dto.jni.MultiFrameAnalysisResponse;
import springbootjni.dto.jni.SpectrumExtractionRequest;
import springbootjni.dto.jni.SpectrumExtractionResponse;
import springbootjni.dto.jni.TriggerCaptureResponse;

import java.util.List;

public interface JNIService {
    BridgeStateResponse getState();

    BridgeStateResponse connect(BridgeConnectRequest request);

    BridgeStateResponse disconnect();

    void sendReset();

    TriggerCaptureResponse sendTriggerOnce(Long userId, boolean autoProcess, String captureScene);

    void sendQueryStatus();

    void sendFullConfig(List<Integer> configBytes);

    void handleImageFrame(int width, int height, short[] pixels16, byte[] pixels8, byte[] fpgaPayload);

    void handleHdrImageFrame(int width, int height, short[] hgPixels16, short[] lgPixels16, byte[] fpgaPayload);

    void handleStatus(int statusBits, int errorCode);

    void handleConfigAck(int resultCode, int failedAddr);

    void handleTransportError(String channel, String message);

    List<ImageFrameResponse> listImages(Long userId);

    List<ImageFrameResponse> listHdrImages(Long userId);

    List<ImageFrameResponse> listHdrDarkImages(Long userId);

    List<ImageFrameResponse> listHdrFlatImages(Long userId);

    ImageFrameResponse processImage(Long userId, long imageId);

    ImagePixelDataResponse getImagePixels(Long userId,
                                          long imageId,
                                          String sourceMode,
                                          String displayFormat,
                                          boolean fullFrame,
                                          Integer xStart,
                                          Integer yStart,
                                          Integer width,
                                          Integer height);

    FpgaPayloadPixelDataResponse getFpgaPayloadPixels(Long userId,
                                                      long imageId,
                                                      Integer start,
                                                      Integer count,
                                                      boolean fullFrame);

    SpectrumExtractionResponse extractSpectrum(Long userId, long imageId, SpectrumExtractionRequest request);

    SpectrumExtractionResponse getLatestSpectrum(Long userId, long imageId);

    boolean deleteImage(Long userId, long imageId);

    int clearImages(Long userId);

    MultiFrameAnalysisResponse analyzeMultiFrame(Long userId, MultiFrameAnalysisRequest request);

    CalibrationSessionResponse generateCalibration(Long userId, String calibrationType, CalibrationRequest request);

    CalibrationSessionResponse buildCalibrationFromImages(Long userId, String calibrationType, CalibrationRequest request);

    List<CalibrationSessionResponse> listCalibrations(Long userId, String calibrationType);

    CalibrationSessionResponse getCalibration(Long userId, long sessionId);

    List<CalibrationPreviewResponse> listCalibrationPreviews(Long userId, long sessionId, int limit);

    CalibrationPreviewResponse getCalibrationReferencePreview(Long userId, long sessionId);

    List<CalibrationPreviewResponse> listCalibrationReferencePreviews(Long userId, long sessionId);

    boolean deleteCalibration(Long userId, long sessionId);

    CalibrationGlobalSettingsResponse getCalibrationGlobalSettings(Long userId);

    CalibrationGlobalSettingsResponse updateCalibrationGlobalSettings(
            Long userId,
            CalibrationGlobalSettingsRequest request);

}
